src/users/keycloak.service.ts

```diff
@@ -0,0 +1,193 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { HttpClientService } from 'src/httpclient/httpclient.service';
+import { KeyCloakConfigs } from './keycloak.configs';
+import { stringify } from 'qs';
+import {
+  Secret,
+  GetPublicKeyOrSecret,
+  verify as verifyCallbackStyle,
+  VerifyOptions,
+} from 'jsonwebtoken';
+import { promisify } from 'util';
+import * as jwkToPem from 'jwk-to-pem';
+import { JWK } from 'jwk-to-pem';
+
+const verify: (
+  token: string,
+  secretOrPublicKey: Secret | GetPublicKeyOrSecret,
+  options?: VerifyOptions,
+) => Promise<Record<string, any> | string> = promisify(verifyCallbackStyle);
+
+interface KeyCloakToken {
+  access_token: string;
+  expires_in: number;
+  refresh_expires_in: number;
+  refresh_token: string;
+  token_type: string;
+}
+
+interface OidcConfig {
+  issuer: string;
+  introspection_endpoint: string;
+  jwks_uri: string;
+  token_endpoint: string;
+}
+
+interface Key {
+  kid: string;
+}
+
+interface Cert {
+  keys: Array<Key & JWK>;
+}
+
+@Injectable()
+export class KeyCloakService {
+  private log: Logger = new Logger(KeyCloakService.name);
+  private keyCloakUsersUri: string;
+  private userName: string;
+  private userPassword: string;
+  private userId: string;
+  private config: OidcConfig;
+  private cert: Cert;
+  private _issuer: string;
+
+  get issuer(): string {
+    return this._issuer;
+  }
+
+  constructor(
+    private readonly configService: ConfigService,
+    private readonly httpClient: HttpClientService,
+  ) {
+    this.loadJwtConfigs();
+
+    this.keyCloakUsersUri = this.configService.get(
+      KeyCloakConfigs.ENV_KEYCLOAK_USERS_URI,
+    );
+    this.userName = this.configService.get(
+      KeyCloakConfigs.ENV_KEYCLOAK_USERS_NAME,
+    );
+    this.userPassword = this.configService.get(
+      KeyCloakConfigs.ENV_KEYCLOAK_USERS_PASSWORD,
+    );
+    this.userId = this.configService.get(KeyCloakConfigs.ENV_KEYCLOAK_USERS_ID);
+  }
+
+  public async getAccessToken(userName: string, password: string) {
+    const data = stringify({
+      password,
+      username: userName,
```

Comment:
Rename `userName` to `username` to be more convenient

```suggestion
      username,
```